File: frontend/src/types/chart.ts
// TypeScript types mirroring backend schemas/chart.py and db.py

export interface CandleBar {
  time: string // "YYYY-MM-DD"
  open: number
  high: number
  low: number
  close: number
}

export interface VolumeBar {
  time: string // "YYYY-MM-DD"
  value: number
}

export interface MAPoint {
  time: string // "YYYY-MM-DD"
  value: number
}

export interface MAOverlay {
  // Daily timeframe fields
  ema10?: MAPoint[]
  ema20?: MAPoint[]
  sma50?: MAPoint[]
  sma100?: MAPoint[]
  sma200?: MAPoint[]
  // Weekly timeframe fields
  sma10?: MAPoint[]
  sma20?: MAPoint[]
  sma40?: MAPoint[]
}

export interface ChartResponse {
  timeframe: string
  candles: CandleBar[]
  volume: VolumeBar[]
  ma: MAOverlay
  rs_line: MAPoint[]
}

// DB update types (mirroring backend schemas/db.py)
export interface UpdateProgress {
  phase: string
  progress: number // 0.0 - 100.0
  current_stock: string | null
  total: number
  eta_seconds: number | null
}

export interface LastUpdated {
  last_updated: string | null
  latest_data_date: string | null // Latest date in stock_prices (YYYY-MM-DD)
  daily_db_size: number // bytes
  weekly_db_size: number // bytes
}

export interface UpdateResult {
  status: string // "completed" | "error"
  success_count: number
  skipped_count: number
  error_count: number
  skipped_codes: string[]
}

File: backend/schemas/chart.py
"""Pydantic schemas for the /api/chart endpoint (TradingView Lightweight Charts format)."""

from __future__ import annotations

from pydantic import BaseModel


class CandleBar(BaseModel):
    """A single OHLC candlestick data point."""

    time: str  # "YYYY-MM-DD"
    open: float
    high: float
    low: float
    close: float


class VolumeBar(BaseModel):
    """A single volume bar data point."""

    time: str  # "YYYY-MM-DD"
    value: float


class MAPoint(BaseModel):
    """A single moving average data point."""

    time: str  # "YYYY-MM-DD"
    value: float


class MAOverlays(BaseModel):
    """Moving average overlays for the chart (all series in TradingView line format).

    Daily timeframe: ema10, ema20, sma50, sma100, sma200 are populated.
    Weekly timeframe: sma10, sma20, sma40 are populated.
    """

    ema10: list[MAPoint] = []
    ema20: list[MAPoint] = []
    sma50: list[MAPoint] = []
    sma100: list[MAPoint] = []
    sma200: list[MAPoint] = []
    sma10: list[MAPoint] = []
    sma20: list[MAPoint] = []
    sma40: list[MAPoint] = []


class ChartResponse(BaseModel):
    """Response payload for GET /api/chart/{code}."""

    timeframe: str = "daily"
    candles: list[CandleBar]
    volume: list[VolumeBar]
    ma: MAOverlays
